### ISSUE TITLE
anyUrl generates invalid JSONPath matcher expression

### ISSUE THREAD

**Issue by pfrank13** [spring-cloud/spring-cloud-contract#250]
This is happening on the 1.0.4.RELEASE, 1.0.5.BUILD-SNAPSHOT and 1.1.0.BUILD-SNAPSHOT version for sure.  In short, if you use anyUrl() in the response body the generated assertion code for junit is valid Java but invalid JSONPath.

Steps to Reproduce
1.  git clone https://github.com/pfrank13/plugin-dependency-scc.git
2. git checkout any_url_bug
3.  ./mvnw clean install -Pextending && ./mvnw clean install
4. observe the stack trace

Output
```java
2017-03-22T11:07:02,808 [main] ERROR com.toomuchcoding.jsonassert.JsonAsserter - Exception occurred while trying to match JSON Path [$[?(@.url =~ /((www\.|(http|https|ftp|news|file)+\:\\/\\/)[_.a-z0-9-]+\.[a-z0-9\\/_:@=.+?,##%&~-]*[^.|\\'|\# |!|\(|?|,| |>|<|;|\)])/)]]
com.jayway.jsonpath.InvalidPathException: Could not find matching close quote for ' when parsing : $[?(@.url =~ /((www\.|(http|https|ftp|news|file)+\:\\/\\/)[_.a-z0-9-]+\.[a-z0-9\\/_:@=.+?,##%&~-]*[^.|\\'|\# |!|\(|?|,| |>|<|;|\)])/)]
	at com.jayway.jsonpath.internal.CharacterIndex.indexOfMatchingCloseChar(CharacterIndex.java:84) ~[json-path-2.2.0.jar:2.2.0]
	at com.jayway.jsonpath.internal.CharacterIndex.indexOfClosingBracket(CharacterIndex.java:113) ~[json-path-2.2.0.jar:2.2.0]
	at com.jayway.jsonpath.internal.path.PathCompiler.readFilterToken(PathCompiler.java:444) ~[json-path-2.2.0.jar:2.2.0]
	at com.jayway.jsonpath.internal.path.PathCompiler.readNextToken(PathCompiler.java:135) ~[json-path-2.2.0.jar:2.2.0]
	at com.jayway.jsonpath.internal.path.PathCompiler.readContextToken(PathCompiler.java:118) ~[json-path-2.2.0.jar:2.2.0]
	at com.jayway.jsonpath.internal.path.PathCompiler.compile(PathCompiler.java:54) ~[json-path-2.2.0.jar:2.2.0]
	at com.jayway.jsonpath.internal.path.PathCompiler.compile(PathCompiler.java:69) ~[json-path-2.2.0.jar:2.2.0]
	at com.jayway.jsonpath.JsonPath.<init>(JsonPath.java:101) ~[json-path-2.2.0.jar:2.2.0]
	at com.jayway.jsonpath.JsonPath.compile(JsonPath.java:467) ~[json-path-2.2.0.jar:2.2.0]
	at com.jayway.jsonpath.internal.JsonContext.read(JsonContext.java:149) ~[json-path-2.2.0.jar:2.2.0]
	at com.jayway.jsonpath.internal.JsonContext.read(JsonContext.java:158) ~[json-path-2.2.0.jar:2.2.0]
	at com.toomuchcoding.jsonassert.JsonAsserter.parseJsonPathAsArray(JsonAsserter.java:205) [jsonassert-0.4.7.jar:?]
	at com.toomuchcoding.jsonassert.JsonAsserter.check(JsonAsserter.java:196) [jsonassert-0.4.7.jar:?]
	at com.toomuchcoding.jsonassert.JsonAsserter.checkBufferedJsonPathString(JsonAsserter.java:213) [jsonassert-0.4.7.jar:?]
	at com.toomuchcoding.jsonassert.JsonAsserter.matches(JsonAsserter.java:159) [jsonassert-0.4.7.jar:?]
	at org.springframework.cloud.contract.verifier.tests.ExampleTest.validate_example(ExampleTest.java:32) [test-classes/:?]
...
```

Generated test line in and: part of the then
```java
DocumentContext parsedJson = JsonPath.parse(response.getBody().asString());
assertThatJson(parsedJson).field("url").matches("((www\\.|(http|https|ftp|news|file)+\\:\\/\\/)[_.a-z0-9-]+\\.[a-z0-9\\/_:@=.+?,##%&~-]*[^.|\\'|\\# |!|\\(|?|,| |>|<|;|\\)])");
```

Contract DSL:
```groovy
Contract.make {
    request {
        method 'GET'
        urlPath '/foo'
        headers {
            accept(applicationJson())
            contentType(applicationJson())
        }
    }
    response {
        status 200
        body ([
            url: anyUrl()
        ])
        headers {
            contentType(applicationJson())
        }
    }
}
```

**Comment by marcingrzejszczak** [comment]
Meh... My bad then. Sorry about this.

**Comment by pfrank13** [comment]
DelegatingJsonVerifiable.escapedHackedJavaText() changes the \\\' into a \' which is then run through the java escaper to be a \\' which is the root of the issue and the malformed json path "Could not find matching close quote for ' when parsing : $[?(@.url =~ /((www\.|(http|https|ftp|news|file)+\:\\/\\/)[_.a-z0-9-]+\.[a-z0-9\\/_:@=.+?,##%&~-]*[^.|\\'|\# |!|\(|?|,| |>|<|;|\)])/)]"

Next step I need to figure out what the JSONPath generated SHOULD be, right now I don't actually know.  I do know simply removing the escapedHackedJavaTest breaks everything =)

**Comment by pfrank13** [comment]
Sadly I'm wrong, I had the order messed up.  I am currently thinking it might be something related not to the code generation because that is actually correct \\' escapes to pattern \' internally which is correct, yet we are seeing \\'.  I think it's whatever escaping is happening during the runtime of the test.

**Comment by pfrank13** [comment]
com.toomuchcoding.jsonassert.JsonAsserter.stringWithEscapedSingleQuotes() sees the ' and escapes it with
```
\'
```
But since it's already escaped we get
```
\\'
```
And that is the bug

**Comment by marcingrzejszczak** [comment]
ha! I think that I fixed it - https://github.com/spring-cloud/spring-cloud-contract/pull/265 can you check out that branch to see if it's satisfactory for you @pfrank13 ? :P

**Comment by pfrank13** [comment]
Verified this works on my end.  

Looking at the fix it appears more of a workaround rather than a fix for regexes that are already escaped though.  Unless I diagnosed the problem incorrectly, but surely you're using a vastly different regex =)

**Comment by marcingrzejszczak** [comment]
Yes and no :P 

The escaping was invalid. For example `hostname` didn't seem to work fine either. So fixing the escaping fixes the problem. OTOH the `url` pattern was not very valid :P I added quite a few tests to check this out. 

Also this PR finally runs the generated code. So I'm 100% sure that the generated test actually passes against the provided JSON. I won't yet merge the PR cause I'll also add some tests by starting a WireMock instance, feeding it with a stub definition and I'll send a request to see if the response is fine. That way we'll have the whole flow checked out.

**Comment by pfrank13** [comment]
Excellent point!  Yeah, I think we're all set here.  Do you usually close these or should I?

**Comment by marcingrzejszczak** [comment]
No, no. It will automatically close once merged to `1.0.x` and `master`. Thanks for checking it out

**Comment by pfrank13** [comment]
No, no, thank you!!!  Glad you're back man, hope you're feeling better =)
